### ISSUE TITLE
Broken with ecto 2.2

### ISSUE THREAD

**Issue by jeffweiss** [jeffweiss/ecto_ldap#19]

**Comment by danielparks** [comment]
Thanks for this issue! It saved me a lot of sadness. I need my sadness for other things.

**Comment by danielparks** [comment]
Hm, actually I ran into the error I was getting before in a different place. I'm on ecto 2.1.6.

### no function clause matching in Ecto.Ldap.Adapter.count_fields/2

~~~
lib/ecto_ldap/adapter.ex
  403  defp ordered_fields(sources) do
  404    {_, model} = elem(sources, 0)
  405    model.__schema__(:fields)
  406  end
  407
* 408  def count_fields(fields, sources) when is_list(fields), do: fields |> Enum.map(fn field -> count_fields(field, sources) end) |> List.flatten
  409  def count_fields({{_, _, fields}, _, _}, sources), do: fields |> extract_field_info(sources)
  410  def count_fields({:&, _, [_idx]} = field, sources), do: extract_field_info(field, sources)
  411
  412  defp extract_field_info({:&, _, [idx]} = field, sources) do
  413    {_source, model} = elem(sources, idx)
~~~

**Comment by OvermindDL1** [comment]
> no function clause matching in Ecto.Ldap.Adapter.count_fields/2

Do you have the rest of the error?  Specifically what arguments were passed to it and what did or did not match?

**Comment by jeffweiss** [comment]
I'll try to take a look at this this weekend or over the holiday break. I recall seeing something at Elixir Conf this fall and thinking, "wow, that would have made ecto_ldap a lot easier," but now I don't remember what it was, only that it was related to function head matching. Brains are the worst.

**Comment by danielparks** [comment]
Sure!

~~~
FunctionClauseError at GET /settings/profile
no function clause matching in Ecto.Ldap.Adapter.count_fields/2
lib/ecto_ldap/adapter.ex
  403  defp ordered_fields(sources) do
  404    {_, model} = elem(sources, 0)
  405    model.__schema__(:fields)
  406  end
  407
➤ 408  def count_fields(fields, sources) when is_list(fields), do: fields |> Enum.map(fn field -> count_fields(field, sources) end) |> List.flatten
  409  def count_fields({{_, _, fields}, _, _}, sources), do: fields |> extract_field_info(sources)
  410  def count_fields({:&, _, [_idx]} = field, sources), do: extract_field_info(field, sources)
  411
  412  defp extract_field_info({:&, _, [idx]} = field, sources) do
  413    {_source, model} = elem(sources, idx)

Ecto.Ldap.Adapter.count_fields/2

Called with 2 arguments
{:{}, [], [{{:., [], [{:&, [], [0]}, :uid]}, [], []}, {{:., [], [{:&, [], [0]}, :cn]}, [], []}]}
{{"users", RobbyWeb.Profile}}

 ecto_ldap lib/ecto_ldap/adapter.ex:408 Ecto.Ldap.Adapter.count_fields/2
 ecto_ldap lib/ecto_ldap/adapter.ex:344 Ecto.Ldap.Adapter.execute/6
 ecto lib/ecto/repo/queryable.ex:130 Ecto.Repo.Queryable.execute/5
 ecto lib/ecto/repo/queryable.ex:35 Ecto.Repo.Queryable.all/4
 ecto lib/ecto/repo/queryable.ex:68 Ecto.Repo.Queryable.one/4
 web/controllers/settings_profile_controller.ex:39 RobbyWeb.SettingsProfileController.show/2
 web/controllers/settings_profile_controller.ex:1 RobbyWeb.SettingsProfileController.action/2
 web/controllers/settings_profile_controller.ex:1 RobbyWeb.SettingsProfileController.phoenix_controller_pipeline/2
 lib/robby_web/endpoint.ex:1 RobbyWeb.Endpoint.instrument/4
 phoenix lib/phoenix/router.ex:278 Phoenix.Router.__call__/1
 lib/robby_web/endpoint.ex:1 RobbyWeb.Endpoint.plug_builder_call/2
 lib/plug/debugger.ex:99 RobbyWeb.Endpoint."call (overridable 3)"/2
 lib/robby_web/endpoint.ex:1 RobbyWeb.Endpoint.call/2
 plug lib/plug/adapters/cowboy/handler.ex:15 Plug.Adapters.Cowboy.Handler.upgrade/4
 cowboy /Users/daniel/work/robby3/deps/cowboy/src/cowboy_protocol.erl:442 :cowboy_protocol.execute/4
~~~

**Comment by danielparks** [comment]
Apologies if this is user error. I don't actually know Elixir.

This was working before, but it's been months and things have bit rotted.

**Comment by OvermindDL1** [comment]
Yeah that argument definitely does not match any of the inputs.  I'd guess it's an error in the definition of the schema or so.  Can you create a [Minimal, Complete, and Verifiable example](https://stackoverflow.com/help/mcve)?  Basically just the minimal definition possible in a project of just ecto and ecto_ldap.  Or at least just copy/paste your entire schema definition here to start (or probably in a new issue)?  :-)

**Comment by danielparks** [comment]
Actually, it looks like I'm on ecto_ldap 0.2.8. Sorry. I've been futzing with versions of all my deps to get this application running again.

**Comment by bossyang** [comment]
I got similar error.
* ecto_ldap 0.3.1
* ecto 2.2.7
* Active Directory

```
iex(1)> Repo.get User, "CN=test,CN=Users,DC=example,DC=com,DC=tw"
** (FunctionClauseError) no function clause matching in Ecto.Ldap.Adapter.count_fields/2

    The following arguments were given to Ecto.Ldap.Adapter.count_fields/2:

        # 1
        %{assocs: [], postprocess: {:from, :any, {:source, :from}},
      preprocess: [{:source, {"users", LdapDemo.User},
        [dn: :string, mail: :string, mobile: :string, sn: :string, cn: :string]}],
      take: []}

        # 2
        {{"users", LdapDemo.User}}

    Attempted function clauses (showing 3 out of 3):

        def count_fields(fields, sources) when is_list(fields)
        def count_fields({{_, _, fields}, _, _}, sources)
        def count_fields({:&, _, [_idx]} = field, sources)

    (ecto_ldap) lib/ecto_ldap/adapter.ex:408: Ecto.Ldap.Adapter.count_fields/2
    (ecto_ldap) lib/ecto_ldap/adapter.ex:344: Ecto.Ldap.Adapter.execute/6
    (ecto) lib/ecto/repo/queryable.ex:133: Ecto.Repo.Queryable.execute/5
    (ecto) lib/ecto/repo/queryable.ex:37: Ecto.Repo.Queryable.all/4
    (ecto) lib/ecto/repo/queryable.ex:70: Ecto.Repo.Queryable.one/4
```

**Comment by jeffweiss** [comment]
Should be fixed via #21 and released as v0.4.0
